arduino-ide-extension/src/browser/theia/core/shell-layout-restorer.ts

```diff
@@ -32,25 +17,165 @@ export class ShellLayoutRestorer extends TheiaShellLayoutRestorer {
+   *   }
+   * ]
+   * ```
+   *
+   * Example with resource only with preview opener:
+   *
+   * ```json
+   * [
+   *   {
+   *     "constructionOptions": {
+   *       "factoryId": "editor-preview-widget",
+   *       "options": {
+   *         "kind": "navigatable",
+   *         "uri": "file:///c%3A/Users/per/Documents/Arduino/duptabs/duptabs.ino",
+   *         "counter": 1,
+   *         "preview": false
+   *       }
+   *     },
+   *     "innerWidgetState": "{\"isPreview\":false,\"editorState\":{\"cursorState\":[{\"inSelectionMode\":false,\"selectionStart\":{\"lineNumber\":1,\"column\":1},\"position\":{\"lineNumber\":1,\"column\":1}}],\"viewState\":{\"scrollLeft\":0,\"firstPosition\":{\"lineNumber\":1,\"column\":1},\"firstPositionDeltaTop\":0},\"contributionsState\":{\"editor.contrib.wordHighlighter\":false,\"editor.contrib.folding\":{\"lineCount\":11,\"provider\":\"indent\"}}}}"
+   *   }
+   * ]
+   * ```
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private filterDescriptions(value: any): WidgetDescription[] {
+    const descriptions = value as WidgetDescription[];
+    const codeEditorUris = new Set<string>();
+    descriptions.forEach(({ constructionOptions }) => {
+      const { options, factoryId } = constructionOptions;
+      if (isResourceWidgetOptions(options)) {
+        const { uri } = options;
+        // resource about to open in code editor
+        if (factoryId === EditorWidgetFactory.ID) {
+          codeEditorUris.add(uri);
+        }
+      }
+    });
+    return descriptions
+      .map((desc) => {
+        const { constructionOptions } = desc;
+        const { options, factoryId } = constructionOptions;
+        if (factoryId === EditorPreviewWidgetFactory.ID) {
+          // resource about to open in preview editor
+          if (isResourceWidgetOptions(options)) {
+            const { uri } = options;
+            // if the resource is about to open in the code editor anyway, do not open the resource in a preview widget too.
+            if (codeEditorUris.has(uri)) {
+              console.log(
+                `Filtered a widget construction options to avoid duplicate editor tab. URI: ${options.uri}, factory ID: ${factoryId}.`
+              );
+              return undefined;
+            } else {
+              // if the preview construction options does not have the code editor counterpart, instead of dropping the preview construction option, a code editor option will be created on the fly.
+              return {
+                constructionOptions: {
+                  factoryId: EditorWidgetFactory.ID,
+                  options: {
+                    uri,
+                    kind: 'navigatable',
+                    counter: 0,
+                  },
+                },
+              };
+            }
+          }
+        }
+        return desc;
+      })
+      .filter(notEmpty);
```

Comment:
for the sake of performance: I see 3 loops here: forEach, map, filter. If we expect this to be time consuming we could think of refactoring it a bit in the future. Btw this is fine as of now